### PR TITLE
Add XMPP Gateway

### PR DESCRIPTION
**Pull request by dora71**
This PR provides XMPP as another gateway for authentication.
By now, only XMPP server prosody with enabled mod_rest or mod_post_msg is supported.

This commit would close issue https://github.com/nextcloud/twofactor_gateway/issues/374

Please review code.

### CODE REVIEW

**Comment by jugendhacker** [comment]
Just my two cents as XMPP user: basing the whole thing on mod_rest makes it compatible with exactly one XMPP server implementation. Okay, yes, Prosody is a quite common setup for public servers, but there are also a lot of servers based on other software.

I would rather suggest implementing something generic based on either BOSH (XEP 0124) or do "real" XMPP via TCP

**Comment by dora71** [comment]
> Just my two cents as XMPP user: basing the whole thing on mod_rest makes it compatible with exactly one XMPP server implementation. Okay, yes, Prosody is a quite common setup for public servers, but there are also a lot of servers based on other software.
> 
> I would rather suggest implementing something generic based on either BOSH (XEP 0124) or do "real" XMPP via TCP

Your idea is principally right. But this was the closeset implementation to what was already been there (Telegram gateway is principally using the same technique), without using new libraries etc. If other XMPP servers have comparable modules, it shouldn't be too difficult to enlarge the gateway to get more servers envolved.

Feel free to improve.

File: lib/Provider/XMPPProvider.php [hunk 1]
```diff
@@ -0,0 +1,69 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * @author Christoph Wurst <christoph@winzerhof-wurst.at>
```
Comment (vitormattos) [inline lib/Provider/XMPPProvider.php:6]:
```suggestion
```

This is a new file, only you by author for now.

File: lib/Service/Gateway/XMPP/Gateway.php [hunk 2]
```diff
@@ -0,0 +1,110 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * @author Christoph Wurst <christoph@winzerhof-wurst.at>
+ * @author André Fondse <andre@hetnetwerk.org>
```
Comment (vitormattos) [inline lib/Service/Gateway/XMPP/Gateway.php:7]:
```suggestion
```
This is a new file, only you by author for now.

File: lib/Service/Gateway/XMPP/GatewayConfig.php [hunk 3]
```diff
@@ -0,0 +1,106 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * @author Christoph Wurst <christoph@winzerhof-wurst.at>
+ * @author André Fondse <andre@hetnetwerk.org>
```
Comment (vitormattos) [inline lib/Service/Gateway/XMPP/GatewayConfig.php:7]:
```suggestion
```
This is a new file, only you by author for now.

File: src/views/XMPPSettings.vue [hunk 4]
```diff
@@ -0,0 +1,38 @@
+<!--
+  - @copyright 2019 Christoph Wurst <christoph@winzerhof-wurst.at>
+  -
+  - @author 2019 Christoph Wurst <christoph@winzerhof-wurst.at>
```
Comment (vitormattos) [inline src/views/XMPPSettings.vue:4]:
```suggestion
  - @copyright 2019 Rainer Dohmen <rdohmen@pensionmoselblick.de>
  -
  - @author 2019 Rainer Dohmen <rdohmen@pensionmoselblick.de>
```

Are you the author

File: tests/Unit/Provider/XMPPProviderTest.php [hunk 5]
```diff
@@ -0,0 +1,107 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * @author Christoph Wurst <christoph@winzerhof-wurst.at>
```
Comment (vitormattos) [inline tests/Unit/Provider/XMPPProviderTest.php:6]:
```suggestion
```

Removing Christoph

**Review by vitormattos — CHANGES_REQUESTED** [review]
Because for now we haven't a test to verify if your implementation is working fine, we will need to believe on your handmade tests.

Sounds ok to me.

Only need to fix the changes at docblocks and the conflict.

**Comment by dora71** [comment]
@vitormattos Hope that everything is alright now.

File: tests/Unit/Provider/XMPPProviderTest.php [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+use OCA\TwoFactorGateway\Provider\XMPPProvider;
+use OCA\TwoFactorGateway\Service\Gateway\XMPP\Gateway;
+use OCA\TwoFactorGateway\Service\StateStorage;
+use OCP\IL10N;
+use OCP\ISession;
+use OCP\Security\ISecureRandom;
+use PHPUnit\Framework\MockObject\MockObject;
+use PHPUnit\Framework\TestCase;
+
+class TelegramProviderTest extends TestCase {
```
Comment (vitormattos) [inline tests/Unit/Provider/XMPPProviderTest.php:36]:
```suggestion
class XMPPProviderTest extends TestCase {
```

**Comment by vitormattos** [comment]
Need to run the command:
`composer cs:fix`

File: lib/Command/Test.php [hunk 7]
```diff
@@ -44,17 +45,22 @@ class Test extends Command {
 	/** @var TelegramGateway */
 	private $telegramGateway;
 
+        /** @var XMPPGateway */
+        private $xmppGateway;
+
```
Comment (vitormattos) [inline lib/Command/Test.php:50]: You will need to run `composer cs:fix` to fix this and other issues that was reported at the CI:

```diff
 > php-cs-fixer fix --dry-run --diff
Loaded config default from "/home/runner/work/twofactor_gateway/twofactor_gateway/.php-cs-fixer.dist.php".
   1) lib/Command/Test.php
      ---------- begin diff ----------
--- /home/runner/work/twofactor_gateway/twofactor_gateway/lib/Command/Test.php
+++ /home/runner/work/twofactor_gateway/twofactor_gateway/lib/Command/Test.php
@@ -45,8 +45,8 @@
 	/** @var TelegramGateway */
 	private $telegramGateway;
 
-        /** @var XMPPGateway */
-        private $xmppGateway;
+	/** @var XMPPGateway */
+	private $xmppGateway;
```

**Review by vitormattos — CHANGES_REQUESTED** [review]

File: lib/Command/Test.php (hunk 7 above)
Comment (dora71) [reply lib/Command/Test.php:50]: Cannot see a difference in the above code (exactly the same lines are removed and then added again).
If I try to do a composer cs:fix as you suggested in my local copy, it stops with an error:
```
> php-cs-fixer fix
sh: 1: php-cs-fixer: not found
```
Commited the change you requested before. Can you run the fix? I am completely new to unit testing.

Comment (vitormattos) [reply lib/Command/Test.php:50]: Was only by example, but wasn't a good example, sorry.
running the command `composer cs:fix` will solve all issues about lint.

Comment (dora71) [reply lib/Command/Test.php]: Finally I got it to work ... hope that now everything is fine and can be merged.

**Review by vitormattos — CHANGES_REQUESTED** [review]
Check the DCO.

You can use `git commit --signof` every time when you do a commit to a Nextcloud app:

Or use the follow commands to create a template to sign-of all commits replacing your name and your email by the right values:

```
git config user.name "Your Name" && git config user.email your@email.tld && git config commit.template /home/username/.git_commit_template
```

And as content of `.git_commit_template you can put:

```


Signed-off-by: Your Name <your@email.tld>
```

But now, to fix the DCO, you will need to run this recommendations:
https://github.com/nextcloud/twofactor_gateway/pull/552/checks?check_run_id=11937823288

**Pull request merged**